### DISCUSSION TITLE
About observing array changes

### DISCUSSION THREAD

**Discussion by vladimirevstratov** [mobxjs/mobx-state-tree#1904] [Q&A] [answered]
What do you think, what is best practice to observe array changes? Because simple replacing array with new array not replacing reference.
Using `.length` and `slice()` in component i think is not the best idea. May be setting `null` before replace array is the best option? Or may be view action in `.views` with returning `arr.slice()` ?

Also i think, that this moment should be crearly described in documentation, because most people stucked on this.

Thanks in advance.

**Comment by EmilTholin** [comment, marked as answer]
Hi @vladimirevstratov!

There is not a silver bullet approach to this problem, as you would most likely approach it differently if it's a small array, if it's a large array, if you only want single elements to re-render on changes, if you are working with a React Native `FlatList`, etc.

Do you have a code example where this is causing you trouble?

**Reply by vladimirevstratov** [reply]
Hi! I will make reproducable example later

**Reply by vladimirevstratov** [reply]
@EmilTholin Hi! 

I have made reproducable example.

If you will click on "Update Array Item" button, author item will change, but! UseEffect will not observe it and will not console log about array changed.
This is a problem with observing arrays in mobx state tree, because reference for arrray items don't change when you change items props.

https://codesandbox.io/s/observing-mobx-array-n3w56k?file=/src/model.ts

**Reply by EmilTholin** [reply]
Hi again @vladimirevstratov!

Ah, now I understand. Yes, the array itself will never change as you said, but you could use [`autorun`](https://mobx.js.org/react-integration.html#useeffect) to make an automatic more fine-grained effect and it will re-run automatically when any observable property you are interested in changes.

**Example**
```ts
useEffect(
  () =>
    autorun(() => {
      console.log(
        "Some book's name was updated",
        rootStore.books.map((b) => b.name)
      );
    }),
  []
);
```

**Reply by vladimirevstratov** [reply]
Thank you, Emil!

And what about React Native `FlatList` and observe array item props changing? 
I've made another example in expo to show you. I found the best solution with wrapping render item return code into `<Observer>` from `mobx-react-lite`.
What do you think about this case?

https://snack.expo.dev/@vladimirevstratov/flatlist-and-mobx-state-tree

Found solution here https://github.com/mobxjs/mobx/issues/1142#issuecomment-323161298

**Reply by vladimirevstratov** [reply]
I think, that all cases should be described in docs of mobx state tree, at least for react native using, because it is very important for big community of developers. 
P.S. it is only my point of view

**Reply by EmilTholin** [reply]
No problem!

Your solution to use `<Observer>` for the `renderItem` component is a nice way to solve it. You could even put it in a separate function outside of the `App` altogether if you prefer, as to not create a new `renderItem` function every time `App` re-renders.

**Example**
```js
const renderAuthor = ({ item }) => (
  <Observer>{() => <View><Text>{item.firstName}</Text></View>}</Observer>
)

const App = observer(() => {
  const { authors } = rootStore;

  return (
    <SafeAreaView>
      <FlatList 
        data={authors}
        renderItem={renderAuthor} />
    </SafeAreaView>
  );
});
```

It's difficult to find the right balance in what to put in the documentation for MST. This is more of a vanilla MobX question that the MobX documentation discusses, and the `FlatList` issue is a very React Native specific problem, but I get what you are saying.

**Reply by vladimirevstratov** [reply]
Thank you 🙏
